File: app/javascript/channels/message_channel.js
import consumer from "channels/consumer" // ここが "channels/consumer" になっていることを確認

document.addEventListener("DOMContentLoaded", () => {
  const sendButton = document.getElementById("send_button");
  const usernameInput = document.getElementById("username_input");
  const messageInput = document.getElementById("message_input");
  const messagesDiv = document.getElementById("messages");

  // メッセージ表示部分の要素が存在することを確認
  if (!messagesDiv) {
    console.error("Messages div not found!");
    return;
  }

  const app = consumer.subscriptions.create("MessageChannel", {
    received(data) {
      // 受信したデータを新しいpタグとして作成し、messagesDivの先頭に追加
      const p = document.createElement("p");
      p.textContent = `${data.name}: ${data.body}`;
      messagesDiv.prepend(p);
    },

    sendMessage(name, msg) {
      // Action Cableチャネルのsend_messageメソッドを呼び出す
      return this.perform("send_message", { name: name, body: msg });
    }
  });

  // 送信ボタンがクリックされた時の処理
  sendButton?.addEventListener("click", (e) => {
    e.preventDefault(); // フォームのデフォルト送信をキャンセル

    const username = usernameInput.value;
    const message = messageInput.value;

    if (username && message) {
      app.sendMessage(username, message);
      messageInput.value = ""; // メッセージ入力フィールドをクリア
    } else {
      alert("Please enter both name and message.");
    }
  }, false);
});

File: app/channels/message_channel.rb
class MessageChannel < ApplicationCable::Channel
  def subscribed
    stream_from "general" # "general" ストリームに接続
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  # クライアントからの `perform("send_message", data)` に対応する
  def send_message(data)
    name = data['name']
    body = data['body']

    # データベースに保存
    message = Message.create(topic: "general", name: name, body: body)

    # 接続している全てのクライアントにメッセージをブロードキャスト
    # received(data) で受け取る形式に合わせる
    ActionCable.server.broadcast("general", { name: message.name, body: message.body })
  end
end
